enhancements/api-review/add-ImagePolicy-to-config.openshift.io-v1alpha1.md

```diff
@@ -0,0 +1,294 @@
+Once an ImagePolicy object gets deleted, container runtime config controller will delete the machine config and run the syncHandler
+
+#### Notes
+
+ImagePolicy is defined as namespaced CRD. `namespace: openshift-config` is used for cluster policy.
+
+### Risks and Mitigations
+
+Risk: The cluster scoped and namespace policies overwrite
+
+Mitigation: MCO container runtime config controller makes sure namespaced policies for an image cannot override global definitions when merging cluster openshift-config policies and namespaced policies into status field of ImagePolicy instances.   
+
+### Drawbacks
+
+## Design Details
+
+### Open Questions [optional]
+
+### Test Plan
+
+**Note:** *Section not required until targeted at a release.*
+
+MCO container runtime config controller can add unit tests and e2e tests.
+- unit test: verify the policies from ImagePolicy instances merged correctly and the controller writes correct format policy.json
+- e2e test: verify the MCO writes configuration to the correct location
+
+
+### Graduation Criteria
+
+#### Dev Preview -> Tech Preview
+
+- Ability to utilize the enhancement end to end
+- End user documentation, relative API stability
+- Sufficient test coverage
+- Gather feedback from users rather than just developers
+- Enumerate service level indicators (SLIs), expose SLIs as metrics
+- Write symptoms-based alerts for the component(s)
+
+#### Tech Preview -> GA
+
+- More testing (upgrade, downgrade, scale)
+- Sufficient time for feedback
+- Available by default
+- Backhaul SLI telemetry
+- Document SLOs for the component
+- Conduct load testing
+- User facing documentation created in [openshift-docs](https://github.com/openshift/openshift-docs/)
+
+**For non-optional features moving to GA, the graduation criteria must include
+end to end tests.**
+
+#### Removing a deprecated feature
+
+Not applicable.
+
+### Upgrade / Downgrade Strategy
+
+Upgrade expectations:
+In order to use ImagePolicy, an existing cluster will have to make on upgrade.
+
+Downgrade expectations: 
+- Not applicable. 
+  If `N` does not support ImagePolicy CRD, `N+1`supports IamgePolicy. It is not expected that the ImagePolicy related to the failure `N->N+1`. New resources should not be created during `N->N+1`.
+
+
+### Version Skew Strategy
+
+The MCO does not require skew check.
+
+### Operational Aspects of API Extensions
+
+- Use-cases that requires over 1000 instances of the CRD is rare, not impacting
+    general API throughput.
+
+#### Failure Modes
+
+- CR field syntax error failure reported from CLI
+- CR conflicting value failure reported by machine config controller logs
+- MCO rolling out configuration file failure reported by MCO
+- If the signature validation fails, then CRI-O will report that in the same way as for any other pull failure. Further enhancements to the kubelet, CRI and CRI-O error handling can be achieved in future Kubernetes releases.
+
+The above errors should not impact the overall cluster health
```

Comment:
(Rolling out a `policy.json` that completely breaks the ability to pull updated images, e.g. for the control plane pods, could be a major problem for a cluster.

Maybe it’s not very different from many other configurations that MCO is managing? Or is it special, and worth some kind of pre-validation?

We do have a special case protecting users against blocking the OCP product repo.)

Reply:
I incline towards preventing the user from setting an ImagePolicy for the OCP product repo. I think the OCP product repos maintained by us are safe enough to use without requiring signature verification?

From the previous [discussion](https://bugzilla.redhat.com/show_bug.cgi?id=1943315#c5), a node drain is required when deleting ICSP to detect image access failure. I am not sure if the node drain& reboot can trigger the CRI-O image pull? If it can trigger the image pull, we can also require the node drain as a pre-validation. But from the feedback of ICSP,  there are customers who request to not have node drain.

Reply:
> I think the OCP product repos maintained by us are safe enough to use without requiring signature verification?

AFAIK OCP doesn’t currently use sigstore; it uses a somewhat-OpenShift-specific variant of simple signing to sign the top-level release image, and all other product images are referenced using digests embedded in that release image.


> I am not sure if the node drain& reboot can trigger the CRI-O image pull?

I think the primary thing to worry is https://github.com/openshift/machine-config-operator/blame/c1a6a342f095553957db9d34aa7881a10b00badf/templates/master/01-master-container-runtime/_base/files/crio.yaml#L6 . But, really, any `policy.json` change (regardless of reboot or image drain) could result in a node that can’t run that control plane pod any more.

(This also links about the discussion about how this is implemented/enforced in CRI-O: does that really happen only o image _pull_, or effectively on every container creation? The latter would make the impact clearly fairly high-risk.

---

WRT possible pre-validation, at least parsing `policy.json`, and possibly evaluating it against a registry source (assuming the registry is reachable!) could be done before actually deploying the configuration. That’s not _too_ much code to potentially run directly inside MCO, though it would be more accurate to run _exactly_ CRI-O’s implementation (to detect cases where MCO is generating a new version of `policy.json` that CRI-O can’t process), and that would require a new CRI-O CLI command or something like that.

Reply:
Addressed that it will not allow configuring a policy for OCP product repo. The validation will parse the scopes to validate that the policy is not for scopes under the OCP product repo.

Reply:
ACK.

Reply:
> Addressed that it will not allow configuring a policy for OCP product repo. The validation will parse the scopes to validate that the policy is not for scopes under the OCP product repo.

Is this in the API anywhere? I didn't see it documented within the API, nor validated, it should be explained to users at the API level that they cannot configure the policy for openshift repositories

Reply:
fixed https://github.com/openshift/enhancements/pull/1402/files#diff-688e8f3b56284c3f0e907836b95cbcb60bcd0676c4819a2e4b1524529bf8d5c5R107